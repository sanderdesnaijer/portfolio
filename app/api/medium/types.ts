export interface MediumArticle {
  author: string;
  categories: string[];
  content: string;
  description: string;
  enclosure: Record<string, unknown>;
  guid: string;
  link: string;
  pubDate: string;
  thumbnail: string | null;
  title: string;
}
