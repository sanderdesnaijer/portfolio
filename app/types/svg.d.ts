// declare module "*.svg" {
//   import * as React from "react";

//   const ReactComponent: React.FunctionComponent<
//     React.SVGProps<SVGSVGElement> & { title?: string }
//   >;

//   export default ReactComponent;
// }

// declare module "*.svg?url" {
//   const content: string;
//   export default content;
// }

// declare module "*.svg?inline" {
//   import * as React from "react";
//   const ReactComponent: React.FunctionComponent<
//     React.SVGProps<SVGSVGElement> & { title?: string }
//   >;
//   export default ReactComponent;
// }
