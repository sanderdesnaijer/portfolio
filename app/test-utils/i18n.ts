import { AbstractIntlMessages } from "next-intl";
import translations from "../translations/en.json";

export type MockTranslations = typeof translations;

export const getTranslationKey = (key: keyof MockTranslations): string => {
  return ((translations as AbstractIntlMessages)[key] || key) as string;
};
