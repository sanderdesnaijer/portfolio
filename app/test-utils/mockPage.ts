import { PageSanity } from "@/sanity/types";

export const mockPage: PageSanity = {
  title: "My page",
  imageAlt: "Profile Picture",
  imageURL: "mock-image-url",
  body: [
    {
      _type: "block",
      children: [
        {
          text: "Mock body content",
          _type: "span",
          marks: [],
        },
      ],
      style: "",
    },
  ],
  name: "",
  description: "",
  slug: {
    _type: "slug",
    current: "page-slug",
  },
  mainImage: {
    _type: "image",
    asset: {
      _ref: "",
      _type: "",
    },
    alt: undefined,
  },
  publishedAt: "",
  _id: "page-1",
  _rev: "",
  _type: "",
  _createdAt: "2025-01-01",
  _updatedAt: "2025-01-02",
};

const mockPage2: PageSanity = {
  title: "My page 2",
  imageAlt: "Page 2 Picture",
  imageURL: "mock-image-url",
  body: [
    {
      _type: "block",
      children: [
        {
          text: "Mock body content page 2",
          _type: "span",
          marks: [],
        },
      ],
      style: "",
    },
  ],
  name: "",
  description: "",
  slug: {
    _type: "slug",
    current: "page2-slug",
  },
  mainImage: {
    _type: "image",
    asset: {
      _ref: "",
      _type: "",
    },
    alt: undefined,
  },
  publishedAt: "",
  _id: "page-2",
  _rev: "",
  _type: "",
  _createdAt: "",
  _updatedAt: "",
};

export const mockPages: PageSanity[] = [mockPage, mockPage2];
