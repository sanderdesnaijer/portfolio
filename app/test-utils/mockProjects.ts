import { ProjectTypeSanity } from "@/sanity/types";

export const mockProject: ProjectTypeSanity = {
  _id: "1",
  title: "Project 1",
  _createdAt: "2025-01-01",
  slug: { _type: "slug", current: "/project-1" },
  imageURL: "http://mocked-image-url.com/image1.jpg",
  mainImage: {
    _type: "image",
    alt: "Project 1 Image Alt",
    asset: { _ref: "string", _type: "ref" },
  },
  body: [
    {
      _type: "block",
      children: [{ text: "Mock body content", _type: "span", marks: [] }],
      style: "",
    },
  ],
  links: [
    { title: "GitHub", icon: "github", link: "http://test" },
    { title: "Live Demo", icon: "external-link", link: "http://test" },
  ],
  publishedAt: "",
  _updatedAt: "",
  _rev: "",
  _type: "",
};

export const mockProjects: ProjectTypeSanity[] = [
  mockProject,
  {
    _id: "2",
    title: "Project 2",
    _createdAt: "2025-01-02",
    slug: { current: "/project-2", _type: "slug" },
    imageURL: "http://mocked-image-url.com/image2.jpg",
    mainImage: {
      _type: "image",
      alt: "Project 2 Image Alt",
      asset: { _ref: "string", _type: "ref" },
    },
    body: null!,
    links: [],
    publishedAt: "",
    _updatedAt: "",
    _rev: "",
    _type: "",
  },
];
