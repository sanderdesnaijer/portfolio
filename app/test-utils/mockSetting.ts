import { SettingSanity } from "@/sanity/types";

export const mockSetting: SettingSanity = {
  _id: "3e37e065-e582-450f-bba0-70709082f578",
  title: "Sander de Snaijer",
  description:
    "Passionate software developer turning creative ideas to functional products.",
  socialMedia: [
    {
      title: "Github",
      link: "https://github.com/sanderdesnaijer",
      icon: "github",
    },
    {
      title: "Linkedin",
      link: "https://www.linkedin.com/in/sanderdesnaijer/",
      icon: "linkedin",
    },
    {
      title: "Gitlab",
      link: "https://gitlab.com/sanderdesnaijer",
      icon: "gitlab",
    },
  ],
  _rev: "",
  _type: "",
  _createdAt: "",
  _updatedAt: "",
  imageURL: "",
  imageAlt: "",
};
