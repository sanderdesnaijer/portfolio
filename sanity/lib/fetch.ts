import "server-only";

import type { QueryParams } from "@sanity/client";
import { draftMode } from "next/headers";
import { client } from "@/sanity/lib/client";
import { REVALIDATE_INTERVAL } from "@/app/utils/constants";

const DEFAULT_PARAMS = {} as QueryParams;
const DEFAULT_TAGS = [] as string[];

export const token = process.env.SANITY_API_READ_TOKEN;

export async function sanityFetch<QueryResponse>({
  query,
  params = DEFAULT_PARAMS,
  tags = DEFAULT_TAGS,
}: {
  query: string;
  params?: QueryParams;
  tags?: string[];
}): Promise<QueryResponse> {
  const currentDraftMode = await draftMode();
  const isDraftMode = currentDraftMode.isEnabled;
  const queryParams = await params;

  if (isDraftMode && !token) {
    throw new Error(
      "The `SANITY_API_READ_TOKEN` environment variable is required."
    );
  }
  const isDevelopment = process.env.NODE_ENV === "development";

  return client
    .withConfig({
      useCdn: isDevelopment,
    })
    .fetch<QueryResponse>(query, queryParams, {
      cache: isDevelopment || isDraftMode ? undefined : "force-cache",
      // cache: isDevelopment || isDraftMode ? undefined : "no-store",
      ...(isDraftMode && {
        token,
        perspective: "previewDrafts",
      }),
      next: {
        revalidate:
          isDevelopment || isDraftMode ? undefined : REVALIDATE_INTERVAL,
        tags,
      },
    });
}
