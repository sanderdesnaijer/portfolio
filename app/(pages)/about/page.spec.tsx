import { render, screen } from "@testing-library/react";
import { sanityFetch } from "@/sanity/lib/fetch";
import Page from "./page";
import { getTranslationKey } from "@/app/test-utils/i18n";
import { JobSanity, PageSanity } from "@/sanity/types";
import { mockPage, mockPages } from "@/app/test-utils/mockPage";
import { mockSetting } from "@/app/test-utils/mockSetting";

jest.mock("next-sanity", () => ({
  ...jest.requireActual("next-sanity"),
  groq: (query: string) => query,

  PortableText: ({ value }: { value: unknown }) => (
    <div data-testid="portable-text">{JSON.stringify(value)}</div>
  ),
}));

describe("app/(pages)/about-me", () => {
  const mockAboutMePage: PageSanity = {
    ...mockPage,
    title: "About Me",
  };

  const mockJobs: JobSanity[] = [
    {
      _id: "job-1",
      imageURL: "mock-job-image-url",
      companyName: "Company A",
      jobTitle: "Developer",
      startDate: "2022-01-01",
      endDate: undefined,
      description: [
        {
          _type: "block",
          children: [
            {
              text: "Job description",
              _type: "span",
              marks: [],
            },
          ],
          style: "",
        },
      ],
      link: "https://test.com",
      tags: [
        {
          _type: "tag",
          name: "React",
          label: "",
          _id: "tag-1",
          _rev: "",
          _createdAt: "",
          _updatedAt: "",
        },
        {
          _type: "tag",
          name: "TypeScript",
          label: "",
          _id: "tag-2",
          _rev: "",
          _createdAt: "",
          _updatedAt: "",
        },
      ],
      _rev: "",
      _type: "",
      _createdAt: "",
      _updatedAt: "",
      employmentType: "full-time",
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the page with correct data", async () => {
    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockAboutMePage)
      .mockResolvedValueOnce(mockJobs)
      .mockResolvedValueOnce(mockSetting)
      .mockResolvedValueOnce(mockPages);

    render(await Page());

    expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
      "About Me"
    );

    expect(
      screen.getByAltText("Profile Picture").getAttribute("src")
    ).toContain("mocked-image-url");

    expect(screen.getByText(JSON.stringify(mockAboutMePage.body))).toBeTruthy();

    expect(screen.getByText("Company A")).toBeInTheDocument();
    expect(screen.getByText("Developer")).toBeInTheDocument();
    expect(
      screen.getByText(
        `Jan 2022 - ${getTranslationKey("pages.about.datePresent")}`
      )
    ).toBeInTheDocument();
  });

  it("renders Page not found when page data is missing", async () => {
    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce(mockJobs)
      .mockResolvedValueOnce(mockSetting)
      .mockResolvedValueOnce(mockPages);

    render(await Page());

    expect(
      screen.getByText(getTranslationKey("error.404.generic.description"))
    ).toBeInTheDocument();
  });

  it("handles missing job links gracefully", async () => {
    const mockJobsWithoutLinks = [
      {
        ...mockJobs[0],
        links: null,
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockAboutMePage)
      .mockResolvedValueOnce(mockJobsWithoutLinks)
      .mockResolvedValueOnce(mockSetting)
      .mockResolvedValueOnce(mockPages);

    render(await Page());

    expect(screen.queryByText("GitHub")).not.toBeInTheDocument();
  });

  it("renders correctly when page body is missing", async () => {
    const mockPageWithoutBody = {
      ...mockAboutMePage,
      body: null,
    };

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockPageWithoutBody)
      .mockResolvedValueOnce([])
      .mockResolvedValueOnce(mockJobs)
      .mockResolvedValueOnce(mockPages);
    render(await Page());

    expect(screen.queryByTestId("portable-text")).not.toBeInTheDocument();

    expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
      "About Me"
    );
  });

  it("renders job with endDate correctly", async () => {
    const mockJobsWithEndDate = [
      {
        ...mockJobs[0],
        endDate: "2023-12-31",
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockAboutMePage)
      .mockResolvedValueOnce(mockJobsWithEndDate)
      .mockResolvedValueOnce(mockJobs)
      .mockResolvedValueOnce(mockPages);

    render(await Page());

    expect(screen.getByText("Jan 2022 - Dec 2023")).toBeInTheDocument();
  });

  it("renders present when job endDate is not given", async () => {
    const mockJobsWithoutEndDate = [
      {
        ...mockJobs[0],
        endDate: null,
      },
    ];

    (sanityFetch as jest.Mock)
      .mockResolvedValueOnce(mockAboutMePage)
      .mockResolvedValueOnce(mockJobsWithoutEndDate)
      .mockResolvedValueOnce(mockJobs)
      .mockResolvedValueOnce(mockPages);

    render(await Page());

    expect(
      screen.getByText(
        `Jan 2022 - ${getTranslationKey("pages.about.datePresent")}`
      )
    ).toBeInTheDocument();
  });
});
