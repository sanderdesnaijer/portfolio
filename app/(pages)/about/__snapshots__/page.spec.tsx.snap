// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/(pages)/about-me renders the page with correct data 1`] = `
<div>
  <main
    class="container mx-auto pt-20 px-4 prose prose-xl dark:prose-invert"
  >
    <h1
      class="font-bold text-8xl mb-10"
    >
      About Me
    </h1>
    <img
      alt="Profile Picture"
      class="absolute right-0 top-0"
      data-nimg="1"
      decoding="async"
      height="300"
      src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75"
      srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=384&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=640&q=75 2x"
      style="color: transparent;"
      width="300"
    />
    <div
      data-testid="portable-text"
    >
      [{"_type":"block","children":[{"text":"Mock body content","_type":"span","marks":[]}],"style":""}]
    </div>
    <h2>
      Experience
    </h2>
    <ol
      class="list-none p-0 flex flex-col gap-10 not-prose"
    >
      <li
        class="flex"
      >
        <div
          class="w-1/5"
        >
          <p
            class="text-base text-right mt-0 mb-2"
          >
            Jan 2022 - present
          </p>
        </div>
        <div
          class="w-4/5 pl-4"
        >
          <div
            class="flex mb-2"
          >
            <img
              alt="mock-job-image-url"
              class="mt-0 h-fit"
              data-nimg="1"
              decoding="async"
              height="48"
              loading="lazy"
              src="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=96&q=75"
              srcset="/_next/image?url=http%3A%2F%2Fmocked-image-url&w=48&q=75 1x, /_next/image?url=http%3A%2F%2Fmocked-image-url&w=96&q=75 2x"
              style="color: transparent;"
              width="48"
            />
            <div
              class="pl-2"
            >
              <h3
                class="text-lg font-bold"
              >
                Company A
              </h3>
              <p
                class="text-base"
              >
                Developer
              </p>
            </div>
          </div>
          <div
            data-testid="portable-text"
          >
            [{"_type":"block","children":[{"text":"Job description","_type":"span","marks":[]}],"style":""}]
          </div>
          <ol>
            <li
              aria-label="github icon"
              class="flex"
              title="github"
            >
              svgrurl
              <div>
                GitHub
              </div>
            </li>
          </ol>
          <p>
            React, TypeScript
          </p>
        </div>
      </li>
    </ol>
  </main>
</div>
`;
